light-prover/prover/insertion_circuit.go

```diff
@@ -0,0 +1,302 @@
+package prover
+
+import (
+	"fmt"
+	"light/light-prover/logging"
+	"math/big"
+	"strconv"
+
+	"github.com/consensys/gnark-crypto/ecc"
+	"github.com/consensys/gnark/backend/groth16"
+	"github.com/consensys/gnark/constraint"
+	"github.com/consensys/gnark/frontend"
+	"github.com/consensys/gnark/frontend/cs/r1cs"
+	"github.com/reilabs/gnark-lean-extractor/v2/abstractor"
+)
+
+type InsertionCircuit struct {
+	// InputHash frontend.Variable `gnark:",public"`
+
+	// private inputs
+	StartIndex frontend.Variable `gnark:"input"`
+	PreRoot    frontend.Variable `gnark:",public"`
+	PostRoot   frontend.Variable `gnark:",public"`
+
+	Leaves       []frontend.Variable   `gnark:"input"`
+	MerkleProofs [][]frontend.Variable `gnark:"input"`
+
+	BatchSize uint32
+	Depth     uint32
+}
+
+func (circuit *InsertionCircuit) Define(api frontend.API) error {
+
+	// api.AssertIsEqual(circuit.InputHash, circuit.InputHash)
+
+	root := abstractor.Call(api, InsertionProof{
+		StartIndex:   circuit.StartIndex,
+		PreRoot:      circuit.PreRoot,
+		Leaves:       circuit.Leaves,
+		MerkleProofs: circuit.MerkleProofs,
+
+		BatchSize: circuit.BatchSize,
+		Depth:     circuit.Depth,
+	})
+	api.AssertIsEqual(root, circuit.PostRoot)
+
+	return nil
+}
+
+type InsertionParameters struct {
+	PreRoot      big.Int
+	PostRoot     big.Int
+	StartIndex   uint32
+	Leaves       []big.Int
+	MerkleProofs [][]big.Int
+}
+
+func (p *InsertionParameters) TreeDepth() uint32 {
+	if len(p.MerkleProofs) == 0 {
+		return 0
+	}
+	return uint32(len(p.MerkleProofs[0]))
+}
+
+func (p *InsertionParameters) BatchSize() uint32 {
+	return uint32(len(p.Leaves))
+}
+
+func (p *InsertionParameters) ValidateShape(treeDepth uint32, batchSize uint32) error {
+	if len(p.Leaves) != int(batchSize) {
+		return fmt.Errorf("wrong number of leaves: %d", len(p.Leaves))
+	}
+	if len(p.MerkleProofs) != int(batchSize) {
+		return fmt.Errorf("wrong number of merkle proofs: %d", len(p.MerkleProofs))
+	}
+	for i, proof := range p.MerkleProofs {
```

Comment:
I don't think it's necessary to check len of `MerkleProofs` during proofgen because we hardcode the len in the circuit as a setup parameter.